### PR TITLE
Pb error msgs

### PR DESCRIPTION
**Pull request by tyeth**
Adds extra logging on encoding errors for protobuf/nanopb. 
Aimed at better logging for better diagnosing of all connectivity / reliability issues, but specifically an issue with pyportal titanos (#585).

Also incorporates PR #589 for RSSI logging to serial (with PR feedback), and PR #590 resetting the esp32 Airlift coprocessor on disconnect which also checks for connectivity if an unsuccessful MQTT ping and triggers reconnect if necessary.

 Lastly adds a macro to "retry a function until timeout with custom retry delay, while verifying return value/func-argument against a condition". Then refactors the networking port specific code to use it in a less delay heavy manner

### CODE REVIEW

**Comment by tyeth** [comment]
Feedback welcomed, just threw this together for that PyPortal Titano user.

**Comment by brentru** [comment]
@tyeth  Have you tested this on physical HW yet and played around with it?

**Comment by tyeth** [comment]
Mainly just ran a few sensors and components on a board or two, didn't see
anything not working

**Review by brentru — APPROVED** [review]

**Comment by tyeth** [comment]
Skipping this release (beta83), will be reviewed in context (PyPortal Titano) next week, then make it into 84.

File: src/network_interfaces/Wippersnapper_AIRLIFT.h [hunk 1]
```diff
@@ -249,25 +287,87 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
...
+      feedWDT();
       WiFi.begin(_ssid, _pass);
       _status = WS_NET_DISCONNECTED;
+
+      // Use the macro to retry the status check until connected / timed out
+      int lastResult;
+      RETRY_FUNCTION_UNTIL_TIMEOUT(
+          []() -> int { return WiFi.status(); }, // Function call each cycle
+          int,                                   // return type
+          lastResult, // return variable (unused here)
```
Comment (tyeth) [inline src/network_interfaces/Wippersnapper_AIRLIFT.h:330]: **not** `(unused here)`

**Comment by tyeth** [comment]
@brentru ready for re-review

File: src/network_interfaces/Wippersnapper_AIRLIFT.h [hunk 2]
```diff
@@ -174,9 +179,34 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
...
+  @returns True if the current version is greater than or
+          equal to the required version, False otherwise.
+  */
+  /********************************************************/
+  bool compareVersions(const String &currentVersion,
+                       const String &requiredVersion) {
+    int curMajor, curMinor, curPatch;
+    int reqMajor, reqMinor, reqPatch;
+
+    sscanf(currentVersion.c_str(), "%d.%d.%d", &curMajor, &curMinor, &curPatch);
```
Comment (brentru) [inline src/network_interfaces/Wippersnapper_AIRLIFT.h:201]: Add handling for if `sscanf` returns an error

File: src/network_interfaces/Wippersnapper_AIRLIFT.h [hunk 3]
```diff
@@ -174,9 +179,34 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
...
+    sscanf(requiredVersion.c_str(), "%d.%d.%d", &reqMajor, &reqMinor,
```
Comment (brentru) [inline src/network_interfaces/Wippersnapper_AIRLIFT.h:202]: Add handling for if `sscanf` returns an error

File: src/network_interfaces/Wippersnapper_WIFININA.h [hunk 4]
```diff
@@ -1,275 +1,284 @@
...
-      @brief  Disconnects from the wireless network.
-  */
-  /**************************************************************************/
-  void _disconnect() {
-    WiFi.disconnect();
-    delay(500);
-  }
-};
-
+/*!
```
Comment (brentru) [inline src/network_interfaces/Wippersnapper_WIFININA.h:1]: Same readability issue as Wippersnapper.h

**Review by brentru — CHANGES_REQUESTED** [review]
@tyeth Overall, this PR is great. It addresses issues when the PB encoder/decoder doesn't work as we expect, airlift-specific issues.

However, it still needs some tweaking before I want to merge. Could you review the following feedback and either comment so we can have a convo about it, or add commits to address them? Thanks!

File: src/network_interfaces/Wippersnapper_AIRLIFT.h [hunk 5]
```diff
@@ -174,9 +179,34 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
...
+  @brief  Compares two version strings.
+  @param  currentVersion
+          Current version string.
+  @param  requiredVersion
+          Required version string.
+  @returns True if the current version is greater than or
+          equal to the required version, False otherwise.
+  */
+  /********************************************************/
+  bool compareVersions(const String &currentVersion,
```
[the remark this answers is not in the document]
Comment (tyeth) [reply src/network_interfaces/Wippersnapper_AIRLIFT.h]: switched back to const char*, been retested on pyportal titano

**Comment by tyeth** [comment]
@brentru ready for re-review, I've dealt with the formatting issue + feedback, retested on pyportal titano.
 Found the RSSI virtual issue. I'd left a breadcrumb of info in another commit message, easily found after walking away from the problem, I just needed to call the method getRSSI unqualified (without WS.) when inside Wippersnapper.cpp/base-class.
 Lastly I was trying to print the RSSI before connect when no longer storing it, so refactored to print in the check_valid_ssid network interface methods instead (where the scan happens and we can briefly get access to all RSSIs before connecting).

**Comment by tyeth** [comment]
Tagging @brentru as ready for re-review

File: src/network_interfaces/Wippersnapper_AIRLIFT.h [hunk 6]
```diff
@@ -234,7 +283,7 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
 protected:
   const char *_ssid;           /*!< Network SSID. */
   const char *_pass;           /*!< Network password. */
-  String _fv;                  /*!< nina-fw firmware version. */
+  const char *_fv;             /*!< nina-fw firmware version. */
```
Comment (brentru) [inline src/network_interfaces/Wippersnapper_AIRLIFT.h:286]: You changed `_fv` from a String (dynamically managed) to a `const char`. I'm not seeing where `_fv` ptr is properly initialized to a value i.e: `const char* _fv = "1.0.0"`

File: src/network_interfaces/Wippersnapper_AIRLIFT.h [hunk 7]
```diff
@@ -249,25 +298,86 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
...
       _status = WS_NET_DISCONNECTED;
+
+      // Use the macro to retry the status check until connected / timed out
+      int lastResult = -1;
+      RETRY_FUNCTION_UNTIL_TIMEOUT(
+          []() -> int { return WiFi.status(); }, // Function call each cycle
+          int,                                   // return type
+          lastResult,                            // return variable
+          [](int status) { return status == WL_CONNECTED; }, // check
+          20000, // timeout interval (ms)
```
Comment (brentru) [inline src/network_interfaces/Wippersnapper_AIRLIFT.h:342]: Make these magic numbers macros

File: src/network_interfaces/Wippersnapper_AIRLIFT.h [hunk 8]
```diff
@@ -249,25 +298,86 @@ class Wippersnapper_AIRLIFT : public Wippersnapper {
...
+          200);  // interval between retries
```
Comment (brentru) [inline src/network_interfaces/Wippersnapper_AIRLIFT.h:343]: Make this a macro

File: src/Wippersnapper.h [hunk 9]
```diff
@@ -59,6 +59,57 @@
...
+            The condition to check the result against.
+    @param  timeout
+            The maximum time to retry the function.
+    @param  interval
+            The time to wait between retries.
+    @param  ...
+            The arguments to pass to the function.
+*/
+/**************************************************************************/
+#define RETRY_FUNCTION_UNTIL_TIMEOUT(func, result_type, result_var, condition, \
```
Comment (brentru) [inline src/Wippersnapper.h:94]: Does this handle negative values?

**Review by brentru — CHANGES_REQUESTED** [review]

**Comment by tyeth** [comment]
@brentru ready for review
Cleaned up the feedback, added the pico cert for staging and wifi101 ignores for esp32 / 8266, and improved the wifi network logging (mentions which SSID it's just connected to).

**Review by brentru — APPROVED** [review]
LGTM! This is a **lot** of changes, do you want to cut a new release for it?

**Comment by tyeth** [comment]
That's a very good idea. Easier roll back that way, but should be fine ☺️

**Pull request merged**